### PR TITLE
It is now possible to choose the edge behavior for tile reflection

### PR DESCRIPTION
**Pull request by Lehonti**
Part of avoiding code repetition involved moving `EdgeBehavior` out of `Warp`, so those files that use `Warp` have changed, too.

Also, the previous samplling functions rounded the positions of the samples, unlike those at `CairoExtensions`, so the output is not 100% the same; which in turn means that the tests had to change.

### CODE REVIEW

File: Pinta.Effects/Effects/TileEffect.cs [hunk 1]
```diff
@@ -166,32 +172,65 @@ private static ColorBgra GetFinalPixelColor (
...
 
-			int adjustedSampleX =
-				(wrappedSampleX < 0)
-				? (wrappedSampleX + settings.size.Width) % settings.size.Width
-				: wrappedSampleX;
-			int adjustedSampleY =
-				(wrappedSampleY < 0)
-				? (wrappedSampleY + settings.size.Height) % settings.size.Height
-				: wrappedSampleY;
+	private static float ReflectCoord (float value, int max)
```
Comment (cameronwhite) [inline Pinta.Effects/Effects/TileEffect.cs:216]: This looks the same as the one in `Warp.cs`, so perhaps it should go in some common location to be reused?
e.g maybe a `GetBilinearSampleReflected()` could make sense

**Review by cameronwhite — COMMENTED** [review]
Nice, this has some good effects 👍 
I noticed the performance seems quite a bit slower running it on a large image, so it might be worth profiling. I suspect it might be due to the new code always doing a bilinear sample?

**Comment by Lehonti** [comment]
Yes @cameronwhite, the performance is suffering because it is always doing a bilinear sample. I think that the performance could be brought back to what it was if we perform quick checks that tell us if that calculation has to be performed or not.

File: Pinta.Effects/Effects/TileEffect.cs (hunk 1 above)
Comment (Lehonti) [reply Pinta.Effects/Effects/TileEffect.cs]: @cameronwhite I just added the extension method you suggested to `CairoExtensions`

Comment (cameronwhite) [reply Pinta.Effects/Effects/TileEffect.cs]: Thanks, that extension method looks good 👍

**Comment by cameronwhite** [comment]
Yeah, previously it was always doing a single sample rounded to integer coordinates, so the main question is whether it's actually a worthwhile quality improvement to do bilinear sampling now (or e.g. whether it's only an improvement when the sample goes beyond the edges and needs wrapping)

**Comment by Lehonti** [comment]
@cameronwhite you have a really good point. I was basically copying what was being done in `Warp`, but it might not be the best idea.

I changed it so that the samples inside the surface are rounded down like before.

**Comment by cameronwhite** [comment]
Yeah I'm fine with this approach - in the future we could have variants of the wrapping / clamping methods which don't do bilinear sampling, if it's not worthwhile for tile reflection. But the performance is much better now so I'm good with this change

**Pull request merged**